### PR TITLE
Add freshness localstore

### PR DESCRIPTION
**Pull request by greatbacon**
This PR adds initial client logic for indicate fresh content in a Space. This is done by comparing a localstore timestamp generated clientside periodically, against the updated `value` on a server side resource.

This PR uses Spaces, but will be replaced with directory entities (once they are added to the session object).

It also does not actually trigger freshening the updated timestamp, that will also happen in a follow up PR.

### CODE REVIEW

File: src/lib/ui/uiMutations.ts [hunk 1]
```diff
@@ -80,6 +83,22 @@ export const ViewSpace: Mutations['ViewSpace'] = async ({
 	}
 };
 
+export const UpdateLastSeen: Mutations['UpdateLastSeen'] = async ({
+	params: {directory_id},
+	ui: {lastSeenByDirectoryId},
+}) => {
+	const $lastSeenByDirectoryId = lastSeenByDirectoryId.get();
```
Comment (ryanatkn) [inline src/lib/ui/uiMutations.ts:90]: no-op suggestion for clarity, use `lastSeenByDirectoryId.mutate` here

File: src/lib/ui/ui.ts [hunk 2]
```diff
@@ -311,6 +315,19 @@ export const toUi = (
...
 				),
 			);
+			lastSeenByDirectoryId.swap(
+				new Map(
+					$session.guest
+						? null
+						: $session.spaces.map(($space) => [
+								$space.directory_id,
+								writable(
+									(browser && localStorage.getItem(`${LAST_SEEN_KEY}${$space.directory_id}`)) ||
```
Comment (ryanatkn) [inline src/lib/ui/ui.ts:325]: something to talk about:

- read a single JSON object or array from storage
- future PR: batch storage writes at the next tick

File: src/lib/ui/SpaceName.svelte [hunk 3]
```diff
@@ -2,11 +2,28 @@
...
+
+	const {
+		ui: {lastSeenByDirectoryId, spaceSelection},
+	} = getApp();
+
+	//TODO replace selected with more cohesive control in Update mutation
+	$: selected = $spaceSelection === space;
+	$: lastSeen = $lastSeenByDirectoryId.value.get($space.directory_id)!;
+	//TODO replace space timestamps with directory timestamps
+	$: systemTime = $space.updated ? $space.updated : $space.created;
```
Comment (ryanatkn) [inline src/lib/ui/SpaceName.svelte:17]: has nice shorthand:

```suggestion
	$: systemTime = $space.updated ?? $space.created;
```

File: src/lib/ui/app.ts [hunk 4]
```diff
@@ -14,6 +14,8 @@ export interface AppStores {
 	devmode: SvelteWritable<boolean>;
 }
 
+export const LAST_SEEN_KEY = 'lastseen:';
```
Comment (ryanatkn) [inline src/lib/ui/app.ts:17]: probably leave a todo to refactor this if we punt to a followup PR, should encapsulate localstorage details in one module

**Review by ryanatkn — APPROVED** [review]

**Review by ryanatkn — APPROVED** [review]

File: src/lib/ui/SpaceName.svelte [hunk 5]
```diff
@@ -2,11 +2,28 @@
...
+	$: clientTime = new Date($lastSeen);
+	$: fresh = !selected && clientTime < systemTime;
 </script>
 
-<SpaceIcon {space} /> <span>{$space.name}</span>
+<SpaceIcon {space} />
+<span
+	>{$space.name}
+	{#if fresh}❗{/if}
+</span>
```
[the remark this answers is not in the document]
Comment (greatbacon) [reply src/lib/ui/SpaceName.svelte:26]: Yeah, visuals aren't final, just wanted something easy to indicate the state. 😁

File: src/lib/ui/ui.ts (hunk 2 above)
Comment (greatbacon) [reply src/lib/ui/ui.ts:325]: I'll add 'em to the list

**Pull request merged**